### PR TITLE
[sdk] Add entry function to existing token SDK

### PR DESCRIPTION
**Pull request by areshand**
### Description

1. adding missing sdk endpoints for all token entry functions, 
2. add a temporary move script to allow user direct call public function (transfer) while we ar e make it an entry function
3. bumped the sdk version

### Test Plan
added unit test

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/5059)

### CODE REVIEW

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { getPropertyValueRaw } from "./property_map_serializer";
+import { bcsSerializeBool, bcsSerializeStr, bcsSerializeU128, bcsSerializeU8, bcsSerializeUint64, Bytes } from "../bcs";
+import assert from "assert";
+
+test("test property_map_serializer", () => {
+  var is_same: (array1: Bytes, array2: Bytes) => boolean = function (array1, array2): boolean {
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts:6]: can we just `function isSame(array1: Bytes, array2: Bytes): boolean {`?

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    return (
+      array1.length == array2.length &&
+      array1.every(function (element, index) {
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts:9]: by convention, we should use an arrow function here. `(element, index) => {`

File: ecosystem/typescript/sdk/src/token_types.ts [hunk 3]
```diff
@@ -46,4 +46,6 @@ export interface Token {
   id: TokenId;
   /** server will return string for u64 */
   amount: U64;
+  /** the property map of the token */
+  token_properties: any;
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/token_types.ts:50]: Can token_properties be defined as `AnyObject` (defined in utils/misc)

File: ecosystem/typescript/sdk/src/token_client.ts [hunk 4]
```diff
@@ -271,6 +342,148 @@ export class TokenClient {
...
+   * @param property_version the version of token PropertyMap
+   * @param amount Amount of tokens which will be transfered
+   * @returns The hash of the transaction submitted to the API
+   */
+  async transferWithOptIn(
+    sender: AptosAccount,
+    creator: MaybeHexString,
+    collectionName: string,
+    tokenName: string,
+    propertyVersion: number,
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/token_client.ts:374]: `BCS.AnyNumber` is better.

File: ecosystem/typescript/sdk/src/token_client.ts [hunk 5]
```diff
@@ -271,6 +342,148 @@ export class TokenClient {
...
+    receiver: MaybeHexString,
+    amount: number,
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/token_client.ts:376]: `BCS.AnyNumber`

File: ecosystem/typescript/sdk/src/move_scripts/token_transfer_with_opt_in/sources/scripts.move [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+        collection_name: String,
+        token_name: String,
+        token_property_version: u64,
+        to: address,
+        amount: u64,
+    ) {
+        let token_id = token::create_token_id_raw(creator, collection_name, token_name, token_property_version);
+        transfer(from, token_id, to, amount);
+    }
+}
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/move_scripts/token_transfer_with_opt_in/sources/scripts.move:17]: what if python SDK wants to call this script as well? Can move this script out of the SDK scope? Also, some comments will be helpful.

Comment (areshand) [reply ecosystem/typescript/sdk/src/move_scripts/token_transfer_with_opt_in/sources/scripts.move:17]: this is temporary. After mainnet release, we should be able to delete this script.

File: ecosystem/typescript/sdk/src/token_client.ts (hunk 5 above)
Comment (areshand) [reply ecosystem/typescript/sdk/src/token_client.ts]: done

File: ecosystem/typescript/sdk/src/token_client.ts (hunk 4 above)
Comment (areshand) [reply ecosystem/typescript/sdk/src/token_client.ts]: done

File: ecosystem/typescript/sdk/src/token_types.ts (hunk 3 above)
Comment (areshand) [reply ecosystem/typescript/sdk/src/token_types.ts]: updated

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts (hunk 2 above)
Comment (areshand) [reply ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts]: updated

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts (hunk 1 above)
Comment (areshand) [reply ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts]: updated

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      const res: AnyNumber = BigInt(values[index]);
+      results.push(bcsSerializeU128(res));
+    } else if (typeTag instanceof TypeTagAddress) {
+      results.push(bcsSerializeStr(values[index]));
+    } else {
+      results.push(bcsSerializeStr(values[index]));
+    }
+  });
+  return results;
+}
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/utils/property_map_serializer.ts:40]: Here is what I meant
```
import {
  Bytes,
  Serializer,
} from "../bcs";
import { TypeTagParser } from "../transaction_builder";
import { serializeArg } from "../transaction_builder/builder_utils";

export function getPropertyValueRaw(values: Array<string>, types: Array<string>): Array<Bytes> {
  if (values.length !== types.length) {
    throw new Error("Invalid input.");
  }

  const results = new Array<Bytes>();
  types.forEach((typ, index) => {
    const typeTag = new TypeTagParser(typ).parseTypeTag();

    const serializer = new Serializer();
    serializeArg(values[index], typeTag, serializer);
    results.push(serializer.getBytes());
  });
  return results;
}

```
Also, we should not use `import assert from "assert"` as it is not working in Browser.  Although we can use it in tests (only node environment), we should use Jest's assertion libraries.

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { getPropertyValueRaw } from "./property_map_serializer";
+import { bcsSerializeBool, bcsSerializeStr, bcsSerializeU128, bcsSerializeU8, bcsSerializeUint64, Bytes } from "../bcs";
+import assert from "assert";
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts:3]: Although assert works in tests, we should use Jest's assertion.

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+      array1.length == array2.length &&
+      array1.every((element, index) => {
+        return element === array2[index];
+      })
+    );
+  }
+  let values = ["false", "10", "18446744073709551615", "340282366920938463463374607431768211455", "hello", "0x1"];
+  let types = ["bool", "u8", "u64", "u128", "0x1::string::String", "address"];
+  let newValues = getPropertyValueRaw(values, types);
+  assert(isSame(newValues[0], bcsSerializeBool(false)));
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts:17]: expect(newValues[0]).toEqual(bcsSerializeBool(false)));

File: ecosystem/typescript/sdk/src/move_scripts/token_transfer_with_opt_in/sources/scripts.move (hunk 6 above)
Comment (jjleng) [reply ecosystem/typescript/sdk/src/move_scripts/token_transfer_with_opt_in/sources/scripts.move:19]: These files will be packaged and get published to NPM. Therefore I still prefer to move this out.

**Comment by jjleng** [comment]
Also, can you please run `yarn lint` and fix the problems?

File: ecosystem/typescript/sdk/src/move_scripts/token_transfer_with_opt_in/sources/scripts.move (hunk 6 above)
Comment (areshand) [reply ecosystem/typescript/sdk/src/move_scripts/token_transfer_with_opt_in/sources/scripts.move:19]: sounds good, I can just keep the bytecode then

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.ts (hunk 7 above)
Comment (areshand) [reply ecosystem/typescript/sdk/src/utils/property_map_serializer.ts:40]: done

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts (hunk 8 above)
Comment (areshand) [reply ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts:3]: done

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts (hunk 9 above)
Comment (areshand) [reply ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts]: try toEqual doesn't work as expect for byte

File: ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  bcsSerializeU128,
+  bcsSerializeU8,
+  bcsSerializeUint64,
+  bcsToBytes,
+  Bytes,
+} from "../bcs";
+import assert from "assert";
+import { AccountAddress } from "../aptos_types";
+import { serialize } from "v8";
+import { from } from "form-data";
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/utils/property_map_serializer.test.ts:14]: this is not used anywhere.

**Review by jjleng — APPROVED** [review]
Please also update changelog

**Review by bowenyang007 — APPROVED** [review]

**Pull request merged**